### PR TITLE
firewall: Make it possible to add custom ports

### PR DESCRIPTION
**Pull request by marusak**
This implements https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/firewall/firewall-dialog-add-service.png (except bottom row in middle, but it provides all needed info, so easy to add as follow up)

It is bigger that I expected it to be, but with all the automatically generated names, id, checking if values make sense etc. it grew a bit :) 
Please no code review yet, I did not check the code after myself, I am opening as I want to discuss the design and usability first.

So @garrett and @andreasn if you have some time could you please check this? Mainly if it works as expected, if there are some things which behave weirdly etc.
And also @garrett I have one favor to ask you (I will happily work in meantime on some of your isses if you need some tests or whatever :) )- if you would have time could you please help me a bit with css in here? Mainly to make it look all nice. I did as much as possible, but the biggest problem I see is that when you open the dialog, search bar is way to wide. Not sure how to make this work in any sensible way.

And as usual some screenshots:
Only one port suggests you description:
onewithlongname
More ports use names or port number if no name available:
more
Invalid values make the input red:
witherrors

### CODE REVIEW

File: pkg/networkmanager/networking.css [hunk 1]
```diff
@@ -198,4 +198,16 @@ h1 .btn-onoff-ct {
...
+.toggle-body > input {
+    margin-right: 2px;
+}
+
+.toggle-body > :not(input) {
+    margin-left: 14px;
+}
+
+.form-control.error {
+    border-color: red;
```
Comment (garrett) [inline pkg/networkmanager/networking.css]: Why not use [native form validation](https://developer.mozilla.org/en-US/docs/Learn/HTML/Forms/Form_validation#The_required_attribute)?

Comment (marusak) [reply pkg/networkmanager/networking.css]: > Why not use native form validation?

Because the checking is rather complex and firstly needs loading `/etc/services`

> Alternatively, there's a .has-error class in PatternFly

awesome, that is what I was looking for :)

Comment (marusak) [reply pkg/networkmanager/networking.css]: > Alternatively, there's a .has-error class in PatternFly, but that needs to be applied to a parent element.

The fact it needs to be applied to a parent element is rather impractical - I need to put \<input\> into \<div\> or other element. We define out own `.has-error` in `pkg/systemd/timer.css`. Would that be a solution?

**Comment by marusak** [comment]
Question @garrett @andreasn 
How the error explaining message should be done? I mean, should there always be space for it or when error occuress jump let all following lines jump? Should the message be smaller and eliminate most of the paddings so the spaces are as small as possible? This to me seems way to flowing around:
text
spaced

**Comment by marusak** [comment]
I've added error message, wrote tests and fixed error that I discovered while writing tests.

**Comment by garrett** [comment]
@marusak: I've ported the services list to form-layout and fixed the alignment of everything else. I've also made some fixes to form-layout along the way to better handle yet-another type of checkbox/radio that we're using here and also an exception to not auto-size spinners.

In addition, I removed a bunch of CSS and added a little bit too, so that the nested form layouts line up as intended here. (Perhaps it could be done better.)

I'm still not happy with the layout of the error messages, as they are spaced to look like they belong to the element below instead of the one above, but that's probably my fault, or my responsibility to fix. I guess it'll have to wait until next week.

Meanwhile, this now probably conflicts with @Gundersanne's service list with ports @ https://github.com/cockpit-project/cockpit/pull/11486. :cry: At least it should be pretty straighforward to fix whichever PR that lands after the other... I hope.

**Comment by garrett** [comment]
Screenshots:

Screenshot_2019-03-28 Sunny

Screenshot_2019-03-28 Sunny(2)

**Comment by garrett** [comment]
I did a quick force-push to fix up the list size on desktop and mobile modes, so it optimizes available space (with a little bit of padding), and readjusted the spinner space to match.

**Comment by marusak** [comment]
> @marusak: I've ported the services list to form-layout and fixed the alignment of everything else. I've also made some fixes to form-layout along the way to better handle yet-another type of checkbox/radio that we're using here and also an exception to not auto-size spinners.

Awesome! Thank you @garrett it looks much better now :)

> I'm still not happy with the layout of the error messages, as they are spaced to look like they belong to the element below instead of the one above, but that's probably my fault, or my responsibility to fix. I guess it'll have to wait until next week.

It was issue even before. I will try to figure out something with it as well.

> Meanwhile, this now probably conflicts with @Gundersanne's service list with ports @ #11486. cry At least it should be pretty straighforward to fix whichever PR that lands after the other... I hope.

Yes, it is conflicting and we are aware of that. But it is nothing we could not solve.

@martinpitt Can I ask you for a fast code review please? It should be fairly close and I don't expect any radical code changes.

File: pkg/networkmanager/firewall-client.js [hunk 2]
```diff
@@ -81,6 +75,23 @@ function initFirewalldDbus() {
...
+
+    firewalld_dbus.subscribe({
+        interface: 'org.fedoraproject.FirewallD1',
+        path: '/org/fedoraproject/FirewallD1',
+        member: 'Reloaded'
+    }, (path, iface, signal, args) => {
+        getServices();
+    });
+
+    getDefaultZonePath()
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.js]: I take it that shouldn't realistically fail in real life. However, could `getDefaultZonePath()` itself at least get a minimal `console.error` catch handler, so that it's easier to debug if this does fail somewhere?

File: pkg/networkmanager/firewall-client.js [hunk 3]
```diff
@@ -81,6 +75,23 @@ function initFirewalldDbus() {
...
+            .then(path => firewalld_dbus.subscribe({
+                interface: 'org.fedoraproject.FirewallD1.config.zone',
+                path: path,
+                member: 'Updated'
+            }, (path, iface, signal, args) => {
+                firewall.reload();
+            }));
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.js]: Please drop the superfluous `{ ; }`. In principle you shouldn't even need a function wrapper at all, just giving `firewall.reload` as the callback should work?

File: pkg/networkmanager/firewall.jsx [hunk 4]
```diff
@@ -114,43 +117,62 @@ class SearchInput extends React.Component {
     render() {
         return <input autoFocus
                       id={this.props.id}
+                      value={this.state.value}
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Curious that this worked before -- with React's [controlled components](https://reactjs.org/docs/forms.html) I thought this pretty much always needs to happen. But this is an unrelated bug fix, can you please split this out into a separate commit?

File: pkg/networkmanager/firewall.jsx [hunk 5]
```diff
@@ -114,43 +117,62 @@ class SearchInput extends React.Component {
...
         };
-
         this.save = this.save.bind(this);
         this.onFilterChanged = this.onFilterChanged.bind(this);
         this.onToggleService = this.onToggleService.bind(this);
+        this.setName = this.setName.bind(this);
+        this.getName = this.getName.bind(this);
+        this.validate = this.validate.bind(this);
+        this.createPorts = this.createPorts.bind(this);
+        this.parse_services = this.parse_services.bind(this);
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Please call this `parseServices()` for consistency.

File: pkg/networkmanager/firewall.jsx [hunk 6]
```diff
@@ -114,43 +117,62 @@ class SearchInput extends React.Component {
...
-                    })
-                }));
+    createPorts() {
+        var ret = [];
+        this.state.custom_tcp_ports.forEach(function(port) {
+            ret.push([port, 'tcp']);
+        });
+        this.state.custom_udp_ports.forEach(function(port) {
+            ret.push([port, 'udp']);
+        });
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Please use arrow function one-liners here.

File: pkg/networkmanager/firewall.jsx [hunk 7]
```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
...
+    parse_services(content) {
+        if (!content) {
+            console.warn("Couldn't read /etc/services");
+            return [ ];
+        }
+
+        var lines = content.split('\n');
+        var ret = {};
+
+        for (var i = 0; i < lines.length; i++) {
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: How about `content.split('\n').forEach(line =>`, so that you avoid the explicit `i` and multiple array access?

File: pkg/networkmanager/firewall.jsx [hunk 8]
```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
         });
     }
 
+    parse_services(content) {
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Can you please add a comment that shows the structure of the returned object?

File: pkg/networkmanager/firewall.jsx [hunk 9]
```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
...
+            return [ ];
+        }
+
+        var lines = content.split('\n');
+        var ret = {};
+
+        for (var i = 0; i < lines.length; i++) {
+            if (!lines[i] || lines[i].startsWith("#"))
+                continue;
+            var m = lines[i].match(/^(\S+)\s+(\d+)\/(\S+).*?(#(.*))?$/);
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Note for self: `services(5)` actually says "The  service-name  should  begin in the first column of the file, since leading spaces are not stripped.", so this is okay.

However, please use `let` or `const` in any block that isn't directly a `function` block (and it's fine to use `let` there as well), due to the broken scoping of `var`.

File: pkg/networkmanager/firewall.jsx [hunk 10]
```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
...
+            port = port.trim();
+            if (!port)
+                return;
+            let p = self.state.avail_services[port.trim()];
+            if (!p) {
+                let num_p = Number(port);
+                if (isNaN(num_p))
+                    error = "Unknown service name";
+                else if (num_p <= 0 || num_p > 65535)
+                    error = "Not valid port number";
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: The error strings need to be i18n'ed. Also, "Invalid port number".

File: pkg/networkmanager/firewall.jsx [hunk 11]
```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
...
+                else
+                    new_ports.push(port);
+            } else if (p['type'].indexOf(targets[0]) < 0)
+                error = "Not matching port number and type";
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: "Mismatching..", or "Port number and type do not match".

Is that an explicit requirement of firewalld? People make up all kinds of ad-hoc ports for their custom services, many of which don't have a `services` entry.

File: pkg/networkmanager/firewall.jsx [hunk 12]
```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
...
+            else {
+                new_ports.push(p.port);
+            }
+        });
+        let newState = { [targets[1]]: new_ports };
+        newState[targets[2]] = error;
+        newState[targets[3]] = event.target.value;
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Any reason to not just initialize it like that with all three keys?

File: pkg/networkmanager/firewall.jsx [hunk 13]
```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
...
+            } else if (all_ports.length > 1) {
+                let max = all_ports.length === 2 ? 2 : 3;
+                for (var i = 0; i < max; i++)
+                    name += this.getName(all_ports[i]) + ", ";
+                if (all_ports.length > 3)
+                    name += "etc.";
+                else
+                    name = name.slice(0, -2);
+            }
+        }
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Please move this out into a helper function. Also, it seems easier to assemble this as a list and use `.join(', ')` in the end. Finally, I'm fairly sure that "etc."  needs to be i18n'ed. How about "…" instead? We use that in several places.

File: pkg/networkmanager/firewall.jsx [hunk 14]
```diff
@@ -181,67 +333,95 @@ class AddServicesBody extends React.Component {
...
+        var addText = this.state.custom ? _("Add Ports") : _("Add Services");
         return (
             <Modal id="add-services-dialog" show onHide={this.props.close}>
                 <Modal.Header>
-                    <Modal.Title> {`Add Services`} </Modal.Title>
+                    <Modal.Title> {addText} </Modal.Title>
                 </Modal.Header>
                 <div id="cockpit_modal_dialog">
-                    {body}
+                    <Modal.Body id="add-services-dialog">
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Can you please move Garrett's commits to the front, and squash this and the last commit? This makes it easier to review, and provides a more logical view on master.

File: test/verify/check-networking-firewall [hunk 15]
```diff
@@ -177,6 +178,66 @@ class TestFirewall(MachineCase):
             b.click("caption button.btn-primary")
             b.wait_present("#cockpit_modal_dialog li input[data-id='empty']")
 
+    @skipImage("Custom ports were introduced in 191", "rhel-7-6-distropkg")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Please refer to the PR #, not a version number -- it's not entirely clear when this will land, and PRs make backporting easier.

File: test/verify/check-networking-firewall [hunk 16]
```diff
@@ -177,6 +178,66 @@ class TestFirewall(MachineCase):
...
+    def testAddCustomServices(self):
+        b = self.browser
+
+        self.login_and_go("/network/firewall")
+
+        # click on the "Add Services" button
+        b.wait_present("caption button.btn-primary:enabled")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Hm, this doesn't actually click. it does in `open_dialog`, but please just move this into that function.

File: test/verify/check-networking-firewall [hunk 17]
```diff
@@ -177,6 +178,66 @@ class TestFirewall(MachineCase):
...
+
+        def open_dialog():
+            b.click("caption button.btn-primary")
+            b.wait_present("#add-services-dialog input[value='ports']")
+            b.click("#add-services-dialog input[value='ports']")
+            b.wait_present("#tcp-ports")
+            b.wait_visible("#tcp-ports")
+
+            time.sleep(1) # We need some time before /etc/services is parsed
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: This needs to be fixed to wait on the effectively visible state change. Maybe the cleanest way would be to disable the button until services is read, or at least disable the "custom" radio button.

File: test/verify/check-networking-firewall [hunk 18]
```diff
@@ -177,6 +178,66 @@ class TestFirewall(MachineCase):
...
+
+        def set_field(sel, val, expected):
+            b.set_input_text(sel, val)
+            b.wait_val("#service-name", expected)
+
+        def check_error(text):
+            b.wait_present(".error-msg:contains({0})".format(text))
+            b.wait_visible(".error-msg:contains({0})".format(text))
+
+        def save(ajde, name, tcp, udp):
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: "ajde"?

**Comment by garrett** [comment]
@marusak: Just checking in quickly on my day off: I'd say we consider this PR done (once it has tests and passes and also passes review), then open a new issue for the validation alignment and I would fix it in the form-layout CSS. It's just a few pixels off above and below; it shouldn't be a big deal.

If you could open that issue and assign it to me, that would be great. Thanks!

File: test/verify/check-networking-firewall [hunk 19]
```diff
@@ -177,6 +178,66 @@ class TestFirewall(MachineCase):
...
+        set_field("#udp-ports", "194,   50000", "http, echo, irc, etc.")
+        set_field("#tcp-ports", "", "irc, 50000")
+        set_field("#tcp-ports", "https", "https, irc, 50000")
+        save("custom--https-irc-50000", "https, irc, 50000", "443", "194, 50000")
+
+        open_dialog()
+        set_field("#tcp-ports", "80", "WorldWideWeb HTTP")
+        b.set_input_text("#service-name", "I am persistent")
+        b.set_input_text("#tcp-ports", "7")
+        time.sleep(5) # We need validate that the service name did not changed
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: This is ugly, but I don't have a better idea off the top of my head. But "need *to* validate … did not change".

**Comment by garrett** [comment]
A request: Please keep the lib/form-layout commits separate. Do not squash them in. I want to make sure it's easy to work on the file (with proper history) and they're independent of the rest of this PR, even though this PR depends on those changes to properly work. (As such, they should probably come first in the order of commits too.) Thanks!

**Comment by andreasn** [comment]
Any idea why it thinks "af" is an OK port? ("a" or "aff" fails as it should)
https://andreasn.portknox.net/s/3xG8jJ2MKekcB7N

**Comment by martinpitt** [comment]
Right, please keep @garrett's commits separate. I was only talking about the first and fifth commit.

@andreasn: "af" is defined in /etc/services, as "AudioFile". I. e. it's fine to give ports by name or by number.

File: test/verify/check-networking-firewall (hunk 18 above)
Comment (marusak) [reply test/verify/check-networking-firewall]: Ah, yeah :smile: 
`id` is built-in function and I prefer not to use it for variables. And this was first thing that I thought of. I will update to something more sensible

File: test/verify/check-networking-firewall (hunk 17 above)
Comment (marusak) [reply test/verify/check-networking-firewall]: I wanted not to disable anything as you still can create custom ports (although only with port numbers). But sure, I can disable radio button until file is read.

File: pkg/networkmanager/firewall.jsx (hunk 13 above)
Comment (marusak) [reply pkg/networkmanager/firewall.jsx]: I am fine with `...` if it can be in the service name (I think it can).

File: pkg/networkmanager/firewall.jsx (hunk 11 above)
Comment (marusak) [reply pkg/networkmanager/firewall.jsx]: This is the one thing I am not entirely sure about. If port is mentioned in /etc/services but not with the selected type if to mark it as error or not. If for example `http - 80 - tcp -  Web Server` would be in /etc/services but not with `udp` what should we do?
1) If 80 is written in UDP
2) If `http` is written in UDP

Number one should be accepted and number two ignored as unknown port name or number two accepted as well and present it as `Web Server`?

File: pkg/networkmanager/firewall.jsx (hunk 4 above)
Comment (marusak) [reply pkg/networkmanager/firewall.jsx]: It worked because parent object is storing the value for itself (and does then filtering) and this component itself did not need to store it explicitly. I needed to introduce it because if you write something into the search bar and then go to `custom ports` and then back the search bar would be empty while the previous value was still filtering. But I will create it as separate commit as it makes sense on its own.

**Comment by marusak** [comment]
Thank you @martinpitt for the review! I will fix all of your comments, also edit commit message and then on next push will let all tests to run.

> Right, please keep @garrett's commits separate. I was only talking about the first and fifth commit.

I wanted to squash his commits but they all made sense as they were. The fifth commit is a bit unlucky, because the part in .jsx should be squashed into the first one, but the .css should not as it is not directly related. So I will probably split that commit into two, the jsx part merge into mine commit and from css will create separate commit.

File: pkg/networkmanager/firewall.jsx (hunk 11 above)
Comment (martinpitt) [reply pkg/networkmanager/firewall.jsx]: Tricky, I don't think there's an obviously right or wrong answer. So perhaps *if* the port is a well-known one, enforce the type, otherwise (for arbitrary unknown ports), allow anything. We can always extend that later on, but this will stay on the safe side.

**Comment by marusak** [comment]
@martinpitt rebased and all comments addressed.

**Comment by marusak** [comment]
Current screenshots also with @Gundersanne changes :)

first
newcustomports

**Comment by marusak** [comment]
Huh, I just noticed on the screenshots that it says "Comma-separated ports, ranges and aliases are accepted" but ranges are not supported (I did not implemented it as I did not realized that I am missing it). I will add it.

**Comment by marusak** [comment]
Pushed version that supports also ranges. Although I suspect ranges would be used only with port number, it is possible to combine aliases and port numbers in any way, so for example `80-https` or `http-https` would work.

**Comment by marusak** [comment]
There are some tests fail. The selected ports do not work well on ubuntu/debian so I need to update that. Also there are some weird warnings such as [these](https://209.132.184.41:8493/logs/pull-11420-20190330-071818-d84c39f9-verify-fedora-atomic/log.html#268) that I cannot reproduce locally.

**Comment by marusak** [comment]
One more problem - It is `<form>` inside of `<form>`, which is not valid html and react is complaning about it. I need to rethink it a bit. @garrett some good idea?

**Comment by garrett** [comment]
> One more problem - It is `<form>` inside of `<form>`, which is not valid html and react is complaning about it. I need to rethink it a bit. @garrett some good idea?

Right. A form cannot have another form.

Here's the current HTML spec:

https://html.spec.whatwg.org/multipage/forms.html#the-form-element

> Content model:
>    Flow content, but with _no form element descendants_.

(Emphasis added)

MDN docs are more approachable, and basically state the same thing:

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form

> Permitted content: Flow content, but _not containing `<form>` elements_

**Comment by garrett** [comment]
@Gundersanne: I have a commit addressing the overflow mentioned @ https://github.com/cockpit-project/cockpit/pull/11486#issuecomment-478163131 and handles it gracefully when there are a bunch of ports. (It tries to put the TCP + UDP side by side, but if it cannot, it wraps.)

It also adds accessibility support by labelling the checkboxes and a few other minor things.

This commit was going to be on its own PR, but it's a bit too intertwined with this PR — the specific commit is: https://github.com/cockpit-project/cockpit/pull/11420/commits/ee21f1d54bb191c92487db6e9ac3aa54f3b1683b

Please review the React changes.

It looks like:

Screenshot_2019-04-01 Sunny(1)

In the screenshot you can see the Satellite entry's ports not fit horizontally, so it wraps the UDP group to a new line. The next entry has room, so both are side-by-side. (I did this using flex and flex-wrap.)

Perhaps I should've commented the CSS a little better before committing?

**Comment by marusak** [comment]
@garrett thank you for the commit :) But it breaks tests. Let me see, I will fix it and force edit your last commit.

**Comment by marusak** [comment]
All related tests pass, @garrett seems (hopefully) happy and therefore ready for re-review. May I ask you @martinpitt ? Thank you :)

File: pkg/networkmanager/firewall.jsx [hunk 20]
```diff
@@ -192,6 +208,165 @@ class AddServicesBody extends React.Component {
...
+                return;
+            let p1, p2;
+            if (port.indexOf("-") > -1) {
+                [p1, p2] = port.split("-");
+                [p1, error] = self.getPortNumber(p1, targets[0], self.state.avail_services[p1]);
+                if (!error) {
+                    [p2, error] = self.getPortNumber(p2, targets[0], self.state.avail_services[p2]);
+                    if (!error) {
+                        if (Number(p1) >= Number(p2))
+                            error = _("Range must strictly ordered");
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: must *be* strictly ordered

File: pkg/networkmanager/firewall.jsx [hunk 21]
```diff
@@ -192,6 +208,165 @@ class AddServicesBody extends React.Component {
...
+
+        let new_ports = [];
+        let self = this;
+        ports.forEach(function(port) {
+            port = port.trim();
+            if (!port)
+                return;
+            let p1, p2;
+            if (port.indexOf("-") > -1) {
+                [p1, p2] = port.split("-");
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: This does not currently handle the case of multiple `-`; e. g.  if you enter `10-50-500`, the `-500` just gets ignored. There should be an error message  instead.

File: pkg/networkmanager/firewall.jsx [hunk 22]
```diff
@@ -210,63 +385,95 @@ class AddServicesBody extends React.Component {
...
+                                </React.Fragment>
+                            }
+                            <div />
+                            <input type="radio" name="type" value="ports" onChange={this.onToggleType} disabled={this.state.avail_services == null} />
+                            {_("Custom Ports")}
+                            { !this.state.custom ||
+                                <fieldset className="ct-form-layout">
+                                    <label className="control-label" htmlFor="hint" hidden>Hint</label>
+                                    <p id="hint">
+                                        {_("Comma-separated ports, ranges and aliases are accepted")}
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Oxford comma: "ranges, and aliases"

File: pkg/networkmanager/firewall.jsx [hunk 23]
```diff
@@ -210,63 +385,95 @@ class AddServicesBody extends React.Component {
...
+                                    </p>
+
+                                    <label className="control-label" htmlFor="tcp-ports">TCP</label>
+                                    <input id="tcp-ports" type="text" onChange={this.validate}
+                                           className={"form-control " + (this.state.tcp_error ? "error" : "") }
+                                           value={this.state.custom_tcp_value}
+                                           placeholder={_("Example: 22,ssh,8080,5900,5910")} />
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: perhaps show the last one as a range, to illustrate what a "range" is? It's otherwise not clear what the separator is. `5900-5910` sounds reasonably realistic, and by themselves they don't introduce a new concept right now.

**Comment by garrett** [comment]
One last thing I thought might be nice: What if the first textarea was focused when the parent radio switcher was changed to it? (The services side of the dialog does this.)

**Comment by marusak** [comment]
@martinpitt updated. @garrett also added focus on the first (tcp) field.
